File: WebApp/indoor-localization/src/modules/report-modules/zone-retention-time-report/managers/AssetIdleTimeStatsManager.ts
import { AssetPositionHistoryLog } from "../../../../entities/AssetPositionHistoryLog";
import AssetIdleTimeCounter from "./AssetIdleTimeCounter";

export default class AssetIdleTimeStatsManager {
  private dataset: AssetPositionHistoryLog[];

  constructor(dataset: AssetPositionHistoryLog[]) {
    this.dataset = dataset;
  }

  private getIdleTimesByAsset(): Record<number, number> {
    const assetGroups: Record<number, AssetPositionHistoryLog[]> = {};
    this.dataset.forEach((log) => {
      if (!assetGroups[log.assetId]) {
        assetGroups[log.assetId] = [];
      }
      assetGroups[log.assetId].push(log);
    });

    const idleTimes: Record<number, number> = {};
    for (const assetId in assetGroups) {
      const counter = new AssetIdleTimeCounter(assetGroups[assetId]);
      idleTimes[Number(assetId)] = counter.getAssetIdleTime();
    }
    return idleTimes;
  }

  public getAvgIdleTime(): number {
    const idleTimes = Object.values(this.getIdleTimesByAsset());
    if (idleTimes.length === 0) return 0;
    return idleTimes.reduce((sum, time) => sum + time, 0) / idleTimes.length;
  }

  public getMinIdleTime(): number {
    const idleTimes = Object.values(this.getIdleTimesByAsset());
    if (idleTimes.length === 0) return 0;
    return Math.min(...idleTimes);
  }

  public getMaxIdleTime(): number {
    const idleTimes = Object.values(this.getIdleTimesByAsset());
    if (idleTimes.length === 0) return 0;
    return Math.max(...idleTimes);
  }
}

File: Android/modules/report_asset_idle_time_module/lib/managers/asset_idle_time_stats_manager.dart
import 'package:indoor_localization/domain/entities/asset_position_history_log.dart';
import 'asset_idle_time_counter.dart';

class AssetIdleTimeStatsManager {
  List<AssetPositionHistoryLog> dataset;

  AssetIdleTimeStatsManager(this.dataset);

  Map<int, double> _getIdleTimesByAsset() {
    final assetGroups = <int, List<AssetPositionHistoryLog>>{};
    for (final log in dataset) {
      if (!assetGroups.containsKey(log.assetId)) {
        assetGroups[log.assetId] = [];
      }
      assetGroups[log.assetId]!.add(log);
    }

    final idleTimes = <int, double>{};
    for (final assetId in assetGroups.keys) {
      final counter = AssetIdleTimeCounter(assetGroups[assetId]!);
      idleTimes[assetId] = counter.getAssetIdleTime();
    }
    return idleTimes;
  }

  double getAvgIdleTime() {
    final idleTimes = _getIdleTimesByAsset().values.toList();
    if (idleTimes.isEmpty) return 0;
    return idleTimes.reduce((sum, time) => sum + time) / idleTimes.length;
  }

  double getMinIdleTime() {
    final idleTimes = _getIdleTimesByAsset().values.toList();
    if (idleTimes.isEmpty) return 0;
    return idleTimes.reduce((a, b) => a < b ? a : b);
  }

  double getMaxIdleTime() {
    final idleTimes = _getIdleTimesByAsset().values.toList();
    if (idleTimes.isEmpty) return 0;
    return idleTimes.reduce((a, b) => a > b ? a : b);
  }
}